Add integration tests for HTTPRoute Accept statuses

mateiidavid opened the pull request:
We currently have no integration tests that create HttpRoutes and assert that the status is set accordingly, based on what `parentReference` they have.

This change adds a set of tests to assert status updates on HttpRoutes function correctly in a k8s environment. Currently, we only test that statuses are set for inbound cases (where a route may reference a `Server`). The three cases we test for:

* That a route has the right condition (accepted) when referencing a `Server` that exists;
* That a route has the right condition (rejected) when referencing a 'Server' that does not exist (in the cluster);
* That the policy controller successfully reconciles a route's status when a new `Server` is created.

To make it easier to test, we also add a util function `await_route_condition` to the testing library.

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

adleong commented on policy-test/src/lib.rs:
```diff
@@ -134,6 +134,30 @@ pub async fn await_pod_ip(client: &kube::Client, ns: &str, name: &str) -> std::n
+    name: &str,
+) -> k8s::policy::httproute::RouteStatus {
+    use k8s::policy::httproute as api;
+    fn get_status(route: &api::HttpRoute) -> Option<api::RouteStatus> {
+        let inner = &route.status.as_ref()?.inner;
+        Some(inner.clone())
+    }
+
+    let route = await_condition(client, ns, name, |obj: Option<&api::HttpRoute>| -> bool {
+        if let Some(route) = obj {
+            return get_status(route).is_some();
+        }
+        false
+    })
+    .await
+    .expect("must fetch route");
+    get_status(&route).expect("route must have a status")
```
consider using less indirection:

```
    let route = await_condition(client, ns, name, |obj: Option<&api::HttpRoute>| -> bool {
        obj.and_then(|route| route.status).is_some()
    })
    .await
    .expect("must fetch route");
    route
        .status
        .map(|s| s.inner)
        .expect("route must have a status")
```

mateiidavid replied:
Good point!

adleong commented on policy-test/tests/accepted_http_route_status.rs:
```diff
@@ -0,0 +1,216 @@
+        .await;
+        // Wait until route is updated with a status
+        let route_status = {
+            let status = await_route_status(&client, &ns, "test-accepted-route").await;
```
any particular reason for this intermediate let binding?  could do:
```
let route_status = await_route_status(&client, &ns, "test-accepted-route")
    .await
    .parents
    .get(0)
    .expect("must have at least one parent status")
    .clone()
```

adleong commented on policy-test/tests/accepted_http_route_status.rs:
```diff
@@ -0,0 +1,216 @@
+        let cond = route_status
+            .conditions
+            .iter()
+            .find(|cond| cond.type_ == "Accepted".to_string())
```
see the clippy suggestion

mateiidavid replied:
Weirdly, clippy suggests dereferencing this. I instead just removed the to_string conversion and it worked.

adleong commented on policy-test/tests/accepted_http_route_status.rs:
```diff
@@ -0,0 +1,216 @@
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn inbound_accepted_reconcile() {
```
is it worth having the opposite test as well where we delete a Server and assert that the accepted status changes from true to false?

mateiidavid replied:
I thought it would be a bit redundant but in hindsight I think it's good to test our index processes delete events accordingly.

adleong reviewed (commented):
Is it worth adding a test for an HTTPRoute with multiple parent refs? (some of which may be accepted and others not?)
Is it worth adding a test for the behavior of parent refs to unknown kinds?
Is it worth adding a test for the behavior when an HTTPRoute has no parent refs?

mateiidavid commented:
Updated: resolved feedback, made some simplifications and added more cases. 

> Is it worth adding a test for an HTTPRoute with multiple parent refs? (some of which may be accepted and others not?)
Is it worth adding a test for the behavior of parent refs to unknown kinds?
Is it worth adding a test for the behavior when an HTTPRoute has no parent refs?

* Multiple parent refs: I've added a test for this. I _think it's worth it_, I was a bit unsure about the semantics here. Services, for example, may only have one route attached to them. I presume we might not support that straight away? Also, it's not a condition that we _need_ for the Server resource, I've added the test but I'm in doubt a bit myself on how we want to proceed here.

* Unknown kinds: can't test this yet :( it will get failed by the admission controller. We should add a test here, similar to the one where there are no parent refs.

* no parent refs: absolutely deserved a test. I wanted to keep the change a bit smaller. I also needed a bit more time to think of a good way to test it. I think what we have in the PR now is good but I'm of course open to suggestions and more than happy to take them into account.

alpeb commented on policy-test/tests/accepted_http_route_status.rs:
```diff
@@ -0,0 +1,421 @@
+        // Status may not be set straight away. To account for that, wrap a
+        // status condition watcher in a timeout.
+        let api = kube::Api::namespaced(client.clone(), &ns);
+        let resolved = tokio::time::timeout(
+            // 5 seconds should be enough for a route to be patched
+            tokio::time::Duration::from_secs(5),
+            kube::runtime::wait::await_condition(
+                api,
+                "test-no-parent-refs-route",
+                |obj: Option<&k8s::policy::HttpRoute>| -> bool {
+                    obj.and_then(|route| route.status.as_ref()).is_some()
+                },
+            ),
+        )
+        .await;
```
Maybe have this timeout logic embedded into `await_route_status` itself, returning an error result if timed out?

mateiidavid replied:
That's a great point. I'm a bit split here... `await_route_status` already uses a timeout since it calls into `await_condition` (which is just a wrapper for the function you see above. https://github.com/linkerd/linkerd2/blob/main/policy-test/src/lib.rs#L71-L75). 

I initially wanted to edit `await_condition` to return the timeout error instead of failing fast once it's been hit, but that meant I had to edit other portions of the code. I ended up keeping it like this because I thought it was simpler (and creating in its own function seemed like too much work).

We could perhaps modify `await_route_status` to instead do all of this manually instead of relying on `await_condition` but then we probably get a bit more duplication. Anyway, given all the context, I'm happy to take a stab at this if you think it's worth resolving. Lmk.

alpeb commented on policy-test/tests/accepted_http_route_status.rs:
```diff
@@ -0,0 +1,421 @@
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn inbound_accepted_no_parent() {
```
Seems like this test can be obviated given the same thing is tested at the beginning of the next one?

mateiidavid replied:
Hmmm, think you're right. Just need to move the asserts 🤔

mateiidavid replied:
Oh, yeah! Hadn't even realised...it does seem to be redundant.

mateiidavid commented on policy-test/tests/accepted_http_route_status.rs:
```diff
@@ -0,0 +1,421 @@
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn inbound_accepted_reconcile_delete_event() {
```
That was my original intent, buut it was simpler to create a new test. I tried to put both of them in the same test and I ended up reverting the changes. Some of the reasons why:

* If we use them both, we have to do the whole clone and dance ritual. Since we move client & namespace into the async blocks we have to clone them prior to it. If we have them together, we have two async blocks (since we need to wait for the status to be updated twice). Not the end of the world but it doesn't add to readability
* I thought the failure point might be a bit more straightforward if the functions test each other's inverse (e.g we know we failed when we went from parent -> no parent).
* Since they execute concurrently we have less waiting time (not that 1m worst-case time is _bad_, we already run the multicluster tests lol)

Most of these are trivial, I think my biggest motivation was that less things happen so imo it's a bit simpler to come back to. I did change the name of the tests though since maybe the `create/delete` doesn't really reflect the behaviour. I'm also okay with making this change, will default to your wisdom here :D

hawkw commented on policy-test/src/lib.rs:
```diff
@@ -61,20 +61,25 @@ pub async fn await_condition<T>(
-    .expect("API call failed")
+    tokio::select! {
+        biased;
+        obj  = kube::runtime::wait::await_condition(api, name, cond) => {
```
tiny whitespace nit (note that `rustfmt` can't format code inside the `select!` macro, sadly...)
```suggestion
        obj = kube::runtime::wait::await_condition(api, name, cond) => {
```

mateiidavid replied:
ughh makes sense that it doesn't but I hadn't realised. Thanks!

hawkw commented on policy-test/tests/accepted_http_route_status.rs:
```diff
@@ -0,0 +1,380 @@
+
+        // Find status for invalid parent and extract the condition
+        let invalid_cond = parent_status
+            .iter()
+            .find(|route_status| route_status.parent_ref.name == "test-invalid-server")
+            .expect("must have a route status set for invalid server")
+            .conditions
+            .clone()
+            .into_iter()
+            .find(|cond| cond.type_ == "Accepted")
```
nit, take it or leave it: it looks like we do this enough that we could _probably_ add a `find_condition` helper (or something) that takes an `impl IntoIter<Item = &RouteStatus>` and a `&str` for the parent ref name and returns an `Option<Condition>`?

hawkw commented on policy-test/tests/accepted_http_route_status.rs:
```diff
@@ -0,0 +1,380 @@
+// whose parentRefs do not exist are patched with an appropriate status.
+async fn inbound_accepted_reconcile_no_parent() {
+    with_temp_ns(|client, ns| async move {
+        // Given a route with inexistent parentReference, we expect to have an
```
nit:
```suggestion
        // Given a route with a nonexistent parentReference, we expect to have an
```

alpeb commented on policy-test/src/lib.rs:
```diff
@@ -61,20 +61,25 @@ pub async fn await_condition<T>(
+    tokio::select! {
+        biased;
+        obj  = kube::runtime::wait::await_condition(api, name, cond) => {
+            let obj = obj.expect("API call failed");
+            Ok(obj)
```
```suggestion
            Ok(obj.expect("API call failed"))
```

alpeb reviewed (approved):
Thanks Matei, LGTM. Just a couple of nits.

mateiidavid commented on policy-test/src/lib.rs:
```diff
@@ -61,20 +61,25 @@ pub async fn await_condition<T>(
+            Ok(obj)
+        },
+
+    _ = time::sleep(time::Duration::from_secs(60)) => {
```
yeah, had some insight in one of the comments above. formatting won't run inside the `select!` macro
